### PR TITLE
New log encoder to report errors through stackdriver

### PR DESCRIPTION
**Pull request by michalderdak**

### CODE REVIEW

File: encoder.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+// The following issue might make this unnecessary:
+// https://github.com/uber-go/zap/issues/514
+type encoder struct {
+	jsonEncoder zapcore.Encoder
+}
+
+// multiline pattern to match the function name line
+var functionNamePattern = regexp.MustCompile(`(?m)^(\S+)$`)
+
+func (s *encoder) EncodeEntry(ent zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
```
Comment (michalderdak) [inline encoder.go:25]: Stolen from https://github.com/evanj/gcplogs/blob/master/gcpzap/encoder.go

File: logger.go [hunk 2]
```diff
@@ -34,43 +33,28 @@ func init() {
...
-		return lvl >= zapcore.ErrorLevel
-	})
-	lowPriority := zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
-		return lvl >= Level.Level() && lvl < zapcore.ErrorLevel
-	})
-
-	// Output channels
-	consoleInfos := zapcore.Lock(os.Stdout)
-	consoleErrors := zapcore.Lock(os.Stderr)
+	err := zap.RegisterEncoder("stackdriver-json", newEncoder)
```
Comment (michalderdak) [inline logger.go:36]: cannot use key "json" as that throws "key already in use" error

File: encoder.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	if ent.Level == zapcore.ErrorLevel && ent.Stack != "" {
```
Comment (lyschoening) [inline encoder.go:26]: Could you explain this? I think we should not pretend something is a panic that isn't. Stackdriver should probably only pick up errors that are actually serious.

File: encoder.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		// Trial-and-error: On App Engine Standard go111 the () are needed after function calls
+		// zap does not add them, so hack it with a regexp
+		replaced := functionNamePattern.ReplaceAllString(ent.Stack, "$1(...)")
+		ent.Message += replaced
+		ent.Stack = ""
+	}
+	return s.jsonEncoder.EncodeEntry(ent, fields)
+}
+
+func (s *encoder) AddArray(key string, marshaler zapcore.ArrayMarshaler) error {
```
Comment (lyschoening) [inline encoder.go:41]: See above.

File: encoder.go (hunk 3 above)
Comment (michalderdak) [reply encoder.go:26]: @lyschoening Nothing is pretending to be a panic. This is a way of getting error stack trace into the correct format so GCP error reporting can pick it up.

File: encoder.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+)
+
+func newEncoder(cfg zapcore.EncoderConfig) (zapcore.Encoder, error) {
+	return &encoder{zapcore.NewJSONEncoder(cfg)}, nil
+}
+
+// Wraps zapcore.Encoder to customize stack traces to be picked up by Stackdriver error reporting.
+// The following issue might make this unnecessary:
+// https://github.com/uber-go/zap/issues/514
+type encoder struct {
```
[the remark this answers is not in the document]
Comment (michalderdak) [reply encoder.go:18]: uu I didn't know, I will try 😄

File: encoder.go (hunk 3 above)
Comment (lyschoening) [reply encoder.go:26]: Alright. Got confused by the comments :)

Comment (michalderdak) [reply encoder.go:26]: @lyschoening  I tried to make it prettier but it seems that GCP error reporting is very sensitive when it comes to stack trace formatting 😄

Comment (lyschoening) [reply encoder.go:26]: It would be nice if we could use the method they expect. They referenced it somewhere. But I guess it would be tricky to reuse.

Comment (michalderdak) [reply encoder.go:26]: @lyschoening they just mention this https://cloud.google.com/error-reporting/reference/rest/v1beta1/projects.events/report?hl=en_GB#reportederrorevent

File: logger.go [hunk 6]
```diff
@@ -18,8 +16,9 @@ var (
 )
 
 type Config struct {
-	Level zapcore.Level `required:"true" default:"warn"`
-	Debug bool          `required:"true" default:"false"`
+	Level       zapcore.Level `required:"true" default:"warn"`
+	Debug       bool          `required:"true" default:"false"`
+	ServiceName string        `required:"true" default:"default_service"`
```
Comment (michalderdak) [inline logger.go:21]: @lyschoening Also I didn't really know how to solve this. If we want to have a nice `service name` in the GCP error reporting we have to specify it (see line 52). So I decided to go with some default value and gradually provide `LOG_SERVICENAME` env variable for every service - that is backwards compatible + seemed like the least number of changes throughout the services

Comment (lyschoening) [reply logger.go:21]: Yes, that should work well for now.

File: encoder.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (michalderdak) [reply encoder.go:18]: @lyschoening I'm still having issues with Go not being object oriented 😅  ... if I inherit `zapcore.Encoder` then I would have to rewrite `EncodeEntry` from scratch as I won't be able to call a parent `EncodeEntry`, I think just calling `EncodeEntry` from within itself creates an infinite recursive call

[the remark this answers is not in the document]
Comment (lyschoening) [reply encoder.go:18]: When you define a base struct like that it also gets embedded as a property. I think you can access it through `s.Encoder.EncodeEntry()`

[the remark this answers is not in the document]
Comment (michalderdak) [reply encoder.go]: Nice! that works

**Review by lyschoening — APPROVED** [review]

**Pull request merged**